### ISSUE TITLE
Symbol is not as hassle-free as claimed

### ISSUE THREAD

**Issue by nagisa** [nagisa/rust_libloading#2]
As noted by comex on [reddit](https://www.reddit.com/r/rust/comments/3s01cz/libloading_a_safer_binding_to_platforms_dynamic/cwtzkan).

**Comment by nagisa** [comment]
Solutions I currently have in mind

### 1.

```
impl<T> Symbol<'a, T> {
    fn forget_lifetime(self) -> UnsafeSymbol<T> { … }
}
```

### 2.

Produce a macro which would basically generate something like

```
struct DynamicMath {
    sin: Symbol<'static, extern fn(f64) -> f64>,
    cos: Symbol<'static, extern fn(f64) -> f64>,
    library: Box<Library>
}
impl DynamicMath {
    fn sin(&self, f64) -> f64 { … }
    fn cos(&self, f64) -> f64 { … }
    fn new(&self, path) -> Result<DynamicMath> { … }
}
```

which can be moved around as one wishes. This would duplicate what `shared_library` does, though, so I’m not sure I want to do this.

### 3.

Leave things as they currently are. The problem’s certainly not new and people could adapt for their own non-trivial uses by transmuting lifetime to `'static` or whatever other solutions are there.

### 4.

Make borrowing the same thing again cheap so calling `get` for same symbol is only potentially expensive the first time.

**Comment by nagisa** [comment]
Something based on https://crates.io/crates/owning_ref could also be possible? That would be some sort of mix between 1 and 4.

**Comment by Tyleo** [comment]
What is the status on this? I would like to use this library for plugins but this is difficult because the symbols have lifetimes.

For this scenario:
1 is unsuitable because the impl must be defined at compile time.
2 is unsuitable because the macro must be defined at compile time.
3 is what I am currently using but requires additional unsafety.
4 is decent but an Rc<Library> in the symbol would probably be cheaper

Why not just provide additional methods which store an Rc to the Library in each symbol? Currently I am doing this in combination with 3 but I feel that the library should take care of this.

**Comment by nagisa** [comment]
> What is the status on this? I would like to use this library for plugins but this is difficult because the symbols have lifetimes.

and

>  4 is decent but an Rc in the symbol would probably be cheaper

I’m still of an opinion that the symbols should not get decoupled from the library they belong to and I’m yet to come across a solution that really satisfies me (e.g. you cannot make anything containing an `Rc` a `Sync+Send`, but using an `Arc` makes the common single-threaded case unnecessarily expensive).

---

I also believe it to be true that one can implement `Rc`/`Arc`/whatever based solution considerably cleanly and in composable manner. For example:

```rust
struct SymbolBundle<T: 'static>(Symbol<'static, T>, Rc<Library>);
impl<T> SymbolBundle<T> {
    unsafe fn new(l: &Rc<Library>, sym: &[u8]) -> SymbolBundle<T> {
        let lc = l.clone();
        let sym: Symbol<T> = l.get(sym).unwrap();
        SymbolBundle(::std::mem::transmute(sym), lc)
    }
}
impl<T> ::std::ops::Deref for SymbolBundle<T> {
    type Target = Symbol<'static, T>;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
```

It also retains a nice property of still abstracting all the unsafety into a single location.

---

>  2 is unsuitable because the macro must be defined at compile time.

Noted.

**Comment by nagisa** [comment]
Another thought that has been floating around in my head lately, is that both the posix and Windows loaders already ref-count the loaded libraries internally, to the best of my knowledge.

Thus an approach like a `Symbol::<'a, T>::detach() -> DetachedSymbol<T>` could be possible, if there was a portable way to duplicate the library handle/increase+decrease the ref-count of the library across the POSIX and Windows’ APIs.

In Windows it seems like the [GetModuleHandleEx](https://msdn.microsoft.com/en-us/library/windows/desktop/ms683200(v=vs.85).aspx) could be used to increase the ref-count for the library.

I can’t find anything in the POSIX API other than `dlopen`-ing the library by the filename again, though. That sounds expensive.

**Comment by nagisa** [comment]
Conversion functions to the unsafe `Symbol` in form of [`Symbol::from_raw`] and [`Symbol::into_raw`] have been implemented. I found myself wanting to improve the experience of using platform-specific stuff easier, which is what those functions accomplish. They also happen to implement a form of option 1 and enable use-cases impossible with the regular tracked `Symbol`. This unsafe `Symbol` could be used to build abstractions as well.

At this point I’m inclined to declare this issue solved, but I feel like some sort of extra abstraction is still desirable.

[`Symbol::from_raw`]: http://nagisa.github.io/rust_libloading/libloading/struct.Symbol.html#method.from_raw
[`Symbol::into_raw`]: http://nagisa.github.io/rust_libloading/libloading/struct.Symbol.html#method.into_raw
